app/activeproject.cpp

```diff
@@ -567,3 +549,49 @@ void ActiveProject::setProjectRole( const QString &role )
     emit projectRoleChanged();
   }
 }
+
+bool ActiveProject::recordingAllowed( QgsMapLayer *layer ) const
+{
+  if ( !layer )
+    return false;
+
+  if ( layer->readOnly() )
+    return false;
+
+  return QgsMapLayerProxyModel::layerMatchesFilters( layer, Qgis::LayerFilter::HasGeometry | Qgis::LayerFilter::WritableLayer ) && layer->id() != positionTrackingLayerId();
+}
+
+QString ActiveProject::positionTrackingLayerId() const
+{
+  if ( !mQgsProject )
+    return QString();
+
+  return mQgsProject->readEntry( QStringLiteral( "Mergin" ), QStringLiteral( "PositionTracking/TrackingLayer" ), QString() );
+}
+
+QList<QgsMapLayer *> ActiveProject::getVisibleLayers() const
+{
+  if ( !mQgsProject )
+    return QList<QgsMapLayer *>();
+
+  QgsLayerTree *root = mQgsProject->layerTreeRoot();
+
+  if ( !root )
+    return QList<QgsMapLayer *>();
+
+  // Get list of all visible and valid layers in the project
+  QList<QgsMapLayer *> visibleLayers;
+  foreach ( QgsLayerTreeLayer *nodeLayer, root->findLayers() )
```

Comment:
I'm used from qgis and dislike the qt macros :)
```suggestion
  const QList<QgsLayerTreeLayer *> nodeLayers = root->findLayers();
  for ( QgsLayerTreeLayer *nodeLayer : nodeLayers )
```